North-Seattle-College/ad440-winter2022-thursday-repo
# Issue 127 Clustering algorithm for comment metadata in Sagemaker

Closes #127 .

## Description
- Via Sagemaker, creates a clustering algorithm using K-Modes to cluster the categorical metadata attributes generated by the `floop-conv-enrich.py` script.
- Implemented via Sagemaker notebook: [iss127-mpg2-clustering-of-fedback-attr](https://us-west-2.console.aws.amazon.com/sagemaker/home?region=us-west-2#/notebook-instances/iss127-mpg2-clustering-of-fedback-attr)

_NOTE: Minor updates made to additional files, per class discussions:_
 - **floop-conv-enrich.py:** Removed the "statement" metadata classification, and grouped it into "sentence".
 - **floop-data-export.py:** Moved from main folder to backend/scripts.
 - **floop-data-import.py:** Moved from main folder to backend/scripts.

## Implementation
- The Sagemaker notebook accesses the enriched floop conversation data json stored in S3 (`enriched-conv-1k-2-sagemaker.json`). Metadata elements for all json entries are pulled an added to an array of dictionaries, which are converted to a pandas dataset.
- An elbow curve algorithm is used to determine optimal number of clusters (which turns out to be 3).
An example of what the Elbow curve graph output might look like:
![image](https://user-images.githubusercontent.com/49133101/157974429-93e2263f-9eda-439e-9c65-3d9fc7f04d48.png)

- Using **KModes clustering** (one of the unsupervised Machine Learning algorithms that is used to cluster categorical variables), we can calculate the dissimilarities between observations and assign them to the closest cluster, as determined by selecting random observations as leaders/cluster definitions. New modes are defined for each cluster, and the process repeats until no re-assignments are required. 
- A "Cluster" column is added to the dataframe, and the determined cluster number is assigned for each metadata set.
- The final dataframe output in Sagemaker will look similar to the below image:
![image](https://user-images.githubusercontent.com/49133101/157557273-5709c195-fb02-455f-a1e8-32b1ee3f980c.png)
- A bar graph is shown to give visual indication of the cluster sizes(similar to the below image):
![image](https://user-images.githubusercontent.com/49133101/157963557-801110be-61de-4049-8a88-17f84c7bd2f7.png)
- The "leaders" for the cluster definitions will be shown in the final cell. These indicate the general attributes of the clusters.

     
## Steps to Verify
1. Login to the AWS [Console](https://aws.amazon.com/console/).
2. Navigate to the [SageMaker landing page](https://us-west-2.console.aws.amazon.com/sagemaker/home?region=us-west-2#/landing).
3. Select **Notebook>Notebook Instances** from the left sidemenu.
4. Find the **iss127-mpg2-clustering-of-fedback-attr** Notebook instance in the list.
5. Select **Open Jupyter Lab** link to the far right of the "iss127-mpg2-clustering-of-fedback-attr" link.
6. Once the Notebook opens, run the cells one at a time.

_NOTE: To restart the notebook to see adjusted results, it's best to select the "restart kernel" icon:_
![image](https://user-images.githubusercontent.com/49133101/157967738-6ffd8e95-d971-4652-a0be-8c87a96ee3ac.png)
_One of the cells adds a column to the dataset called "Cluster", and if this cell ends up being run more than once, it will add multiple "Cluster" columns and throw off some subsequent formulas._

7. Note the Dataframe output listed above, the bar graph showing cluster distribution, and the Cluster definition leaders output.
